lib/ClangImporter/ClangImporter.cpp

```diff
@@ -2743,6 +2743,36 @@ void ClangImporter::lookupRelatedEntity(
   }
 }
 
+NominalTypeDecl *
+ClangImporter::instantiateCXXClassTemplate(
+    clang::ClassTemplateDecl *decl,
+    ArrayRef<clang::TemplateArgument> arguments) {
+  void *InsertPos = nullptr;
+  auto *ctsd = decl->findSpecialization(arguments, InsertPos);
+  if (!ctsd) {
+    ctsd = clang::ClassTemplateSpecializationDecl::Create(
+        decl->getASTContext(), decl->getTemplatedDecl()->getTagKind(),
+        decl->getDeclContext(), decl->getTemplatedDecl()->getBeginLoc(),
+        decl->getLocation(), decl, arguments, nullptr);
+    decl->AddSpecialization(ctsd, InsertPos);
+  }
+
+  auto CanonType = decl->getASTContext().getTypeDeclType(ctsd);
+  assert(isa<clang::RecordType>(CanonType) &&
+          "type of non-dependent specialization is not a RecordType");
```

Comment:
Could this happen because of an internal error (this check is already handled elsewhere but maybe there's a bug there) or is this potentially because of a user error? If it could be because of a user error, please at least add a TODO for surfacing a diagnostic here.

Reply:
I'm closely following the logic in Clang (https://github.com/llvm/llvm-project/blob/1abb883a048153c83a4e11070219d23f362e7377/clang/lib/Sema/SemaTemplate.cpp#L3651), so it should be a programmer error when this hits, not a user error. I'm happy to remove the assertion if you think it doesn't make sense to have it.

Reply:
(Sorry, missed this earlier.)

If Clang also does the same thing, I think it's fine. 👍🏼